sklearn/ensemble/stacking.py

```diff
@@ -0,0 +1,221 @@
+            if getattr(self.base_estimator, 'predict_proba', None):
+                method = 'predict_proba'
+            elif getattr(self.base_estimator, 'decision_function', None):
+                method = 'decision_function'
+            else:
+                method = 'predict'
+        else:
+            method = self.method
+
+        return method
+
+    def transform(self, *args, **kwargs):
+        """Transform dataset.
+
+        Note that, unlike ``fit_transform()``, this won't return the cross
+        validation predictions. See the :ref:`User Guide <ensemble>` for more
+        info.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix, shape=(n_samples, n_features)
+            Input data to be transformed. Use ``dtype=np.float32`` for maximum
+            efficiency. Sparse matrices are also supported, use sparse
+            ``csr_matrix`` for maximum efficiency.
+
+        Returns
+        -------
+        X_transformed : sparse matrix, shape=(n_samples, n_out)
+            Transformed dataset.
+
+        """
+        t = getattr(self.base_estimator, self._method_name())
+        preds = t(*args, **kwargs)
+
+        if preds.ndim == 1:
+            preds = preds.reshape(-1, 1)
+
+        return preds
+
+    def fit_transform(self, X, y, **fit_params):
+        """Fit estimator and transform dataset.
+
+        Note that this behavior is different from ``fit().transform()`` as it
+        will return the cross validation predictions instead.. See the
+        :ref:`User Guide <ensemble>` for more info.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix, shape=(n_samples, n_features)
+            Input data used to build forests. Use ``dtype=np.float32`` for
+            maximum efficiency.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        **fit_params : parameters to be passed to the base estimator.
+
+        Returns
+        -------
+        X_transformed : sparse matrix, shape=(n_samples, n_out)
+            Transformed dataset.
+        """
+        preds = cross_val_predict(self.base_estimator, X, y, cv=self.cv,
+                                  method=self._method_name(),
+                                  n_jobs=self.n_jobs, fit_params=fit_params)
+
+        self.base_estimator.fit(X, y, **fit_params)
+
+        if preds.ndim == 1:
+            preds = preds.reshape(-1, 1)
+
+        return preds
+
+
+def _identity(x):
+    return x
+
+
+def _identity_transformer():
+    """Contructs a transformer that returns its input unchanged"""
+    return FunctionTransformer(_identity, accept_sparse=True)
+
+
+def make_stack_layer(base_estimators, restack=False, cv=3, method='auto',
+                     n_jobs=1, n_cv_jobs=1, transformer_weights=None):
+    """ Construct single layer for model stacking
+
+    Parameters
+    ----------
+    base_estimators : list of estimators to be used in stacking
```

Comment:
description should be on next line.

this is actually expected to be a list of (name, estimator) pairs.

I'm not convinced that we want it to be a list of (name, estimator) pairs; compatibility with `make_union` makes some sense given the function name.

Reply:
it does make sense, but then again, we don't have the `FeatureUnion` counterpart to expose a way to set explicit names for estimators

Reply:
a way of solving it would be to allow `FeatureUnion`, `Pipeline`  (and consequently `make_stack_estimator`) to accept both tuples of `name, estimator` and `estimator` alone. it would then name the ones that need to be named. this way `make_feature_union` and `make_pipeline` wouldn't even be needed. if you think it's a good idea, I can implement it on another PR

Reply:
I don't think so. That's the sort of thing we could consider once `Pipeline` properly adhered to the convention that it should not modify input parameters. But for legacy reasons, `Pipeline` is different and modifies `steps`. Without fixing as in #8350, I don't think we can allow both.

> it does make sense, but then again, we don't have the FeatureUnion counterpart to expose a way to set explicit names for estimators

No, we don't, but that's because `make_stack_layer` is a fairly thin wrapper around `FeatureUnion` already.